*: Data collection from FCOS machines

zonggen opened the pull request:
Source of truth for each part of data collected:
- Minimal:
  - platform: read_file `/proc/cmdline`
  - current_os_version: read_output `rpm-ostree status --json`
  - original_os_version: read_file `/.coreos-aleph-version.json`
  - instance_type: read_file `/run/metadata/afterburn`
- Full:
  - hardware:
    - disk: read_output `lsblk --fs --json`
    - cpu: read_output `lscpu --json`
    - memory: read_output `lsmem --json`
  - network: read_output `nmcli device show`
  - container_runtime:
      - podman: 
          - check if running by `pgrep podman`
          - count running container by `pgrep conmon | wc -l`
     - docker: 
          - check if running by `pgrep dockerd`
          - count running container by `pgrep containerd-shim | wc -l`
      - cri-o: 
          - check if running by `pgrep crio`
          - count running container by `pgrep crictl | wc -l`
      - system-nspawn: 
          - check if running and counting containers by `pgrep systemd-nspawn`

Signed-off-by: Allen Bai, abai@redhat.com

Commit:
src/identity: Add a module that collects platform id

Collects ignition platform id from /proc/cmdline in the varriable
`ignition.platform.id=` and stores the collected data inside the
`identity::Identity` struct.

Also adds different grains of data collection level: minimum and
full, depending on the passed toml configuration.

Signed-off-by: Allen Bai, abai@redhat.com

zonggen commented:
Currently in `main.rs` the collected `ignition platform id` will be printed to the console, which will be updated later to be bundled with other data and sent to the remote endpoint.

zonggen commented:
Also, this PR will be updated to resolve https://github.com/coreos/fedora-coreos-pinger/issues/30#issue-494138139

lucab commented on src/main.rs:
```diff
@@ -24,6 +29,15 @@ fn check_config(config: inputs::ConfigInput) -> failure::Fallible<()> {
     Ok(())
 }
 
+fn send_metrics(id: &identity::Identity) -> failure::Fallible<()> {
```
We are trying not to call these "metrics", as it just creates confusion with what most of the people call "metrics", i.e. https://prometheus.io/docs/concepts/metric_types/.

Unless we are planning to actually collect and push only that kind of `(string, float)` records (is that the case?), let's try to find a better name for this.

zonggen replied:
No, that's not the case. Will be renaming those bits and other parts with the name "metrics". Thanks for the clarification!

zonggen replied:
Updated

lucab commented on src/identity/platform.rs:
```diff
@@ -0,0 +1,97 @@
+//! Kernel cmdline parsing - utility functions
+//!
+//! NOTE(lucab): this is not a complete/correct cmdline parser, as it implements
```
At some point I guess I'll start bribing @jlebon to turn this into a proper library, but today is not that day :)

zonggen replied:
Totally agreed! I feel like a lot of code will be similar between `Zincati`/`Afterburn` and `fedora-coreos-pinger`. Like you said, today is not the day ;)

lucab commented on src/identity/mod.rs:
```diff
@@ -0,0 +1,111 @@
+    pub fn try_default(level: &str) -> Fallible<Self> {
+        let platform = platform::read_id(KERNEL_ARGS_FILE)?;
+
+        let id = match level {
```
I think you can factor out a bunch of stuff from here, like this:
```rust
let level = match input {
  "minimal" | "full" => input.to_string(),
 _ => "minimal".to_string(),
};
```

zonggen replied:
Yes, that should be cleaner :+1:

lucab commented on src/identity/mod.rs:
```diff
@@ -0,0 +1,111 @@
+
+    /// Getter for collected metrics, returned as a HashMap
+    pub fn get_metrics(&self) -> HashMap<String, String> {
+        let mut vars = HashMap::new();
```
In case you don't know it, check https://docs.rs/maplit/1.0.2/maplit/macro.hashmap.html. It could be handy here I guess.

zonggen replied:
Ah, that's better!

Commit:
*: Update according to reviews

Changes the use of term "metrics", deduplicates match pattern and
uses maplit::hashmap! macro.

Commit:
identity/os-release: Add a module to abstract os version

zonggen commented:
UPDATE:
 - Add the module `os-release.rs` under identity

_Note_: Most portion of code is the same as `platform.rs` so in the future, towards the end of this PR should de-duplicate this part of code.

Commit:
identity/os_release.rs: Collect CoreOS alpha version

Picks up CoreOS alpha version information from `/.coreos-aleph-version.json`.

Signed-off-by: Allen Bai, abai@redhat.com

jlebon commented on src/identity/mod.rs:
```diff
@@ -0,0 +1,148 @@
+/// OS release file location
+static OS_RELEASE_FILE: &str = "/etc/os-release";
+/// OS alpha version file
+static OS_ALPHA_VERSION_FILE: &str = "/.coreos-aleph-version.json";
```
Did you mean `OS_ALEPH_VERSION_FILE` here and elsewhere?
Not that ALPHA is terribly wrong, but "CoreOS Alpha version" is kinda confusing already with the fact that CL has an Alpha channel. :)

zonggen replied:
Ahh.. Right, it should be `OS_ALEPH_VERSION_FILE`. Don't know why my brain interpret to `Alpha` all the time ..

jlebon commented on src/identity/os_release.rs:
```diff
@@ -0,0 +1,114 @@
+
+}
+
+/// Read current os version info from os release file.
```
This is OK, though note the source of truth is the OSTree commit metadata. The `os-release` file is mutated using a treefile option, and e.g. if that option stops working for some reason ([as it has in the past](https://github.com/projectatomic/rpm-ostree/pull/1481)) and we somehow ship it, this code will fail.

The easy way to do this is to just do like Zincati and call `rpm-ostree status --json` and parse that. The harder (but arguably cleaner) way is to e.g. use the [ostree](https://docs.rs/ostree) and [glib](https://docs.rs/glib) crates to do this. (Though from quickly browsing [the glib docs](https://docs.rs/glib/0.8.1/glib/variant/struct.Variant.html), I'm not sure if all the bindings are there to get at the commit metadata tuple from the commit variant object).

Doesn't have to be done here though, feel free to just file an issue for now if you prefer!

zonggen replied:
I see, will be using `rpm-ostree status --json` for now
But the latter option sounds better though, would be great if there's rpm-ostree crate for rust

jlebon reviewed (commented):
Looks sane from a quick scan!

Commit:
src/providers: Extract cloud instance type in mod `providers`

Commit:
os_release.rs: Change name alpha to aleph

Commit:
rpm-ostree: Add rpm-ostree module for os version extraction

Commit:
*: cleanup redundant code imported from Afterburn

Commit:
src/agent: merge minimal and full under the same module, agent
src/full: add hardware data collection

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
minimal/mock_tests: add mock test for minimal agent

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
full/mock_tests: add mock tests for full agent

Signed-off-by: Allen Bai <abai@redhat.com>

zonggen commented:
STATUS:
Currently Travis CI is failing since it is using Rust docker image which is based on Debian and it does not ship with `lsmem` and `lscpu` does not contain an `--json` option.
To test with more production-like environment, i.e. FCOS, it might be better to switch to UBI (https://www.redhat.com/en/blog/introducing-red-hat-universal-base-image). Will be updating with changes to see if it works.

Commit:
full/lsmem: allow "yes"/"no" for field `removable`
.travis.yml: test inside ubi container instead of ubuntu vm

Allows "yes"/"no" values for `removable` field of `lsmem --json`, because of
different implementations, the field type might be different. However, in an
FCOS machine this will be true/false. This change is mainly to accommodate testing
environment.

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
src: format with rustfmt

Signed-off-by: Allen Bai <abai@redhat.com>

agent/full: only collect hw info on bare metals

Signed-off-by: Allen Bai <abai@redhat.com>

full/network: collect network info with `nmcli`

Collects network data with format `[key[\s]+value\n]+`,
to see example run `nmcli device show`. Parse the data
as key:value pairs stored in a HashMap.

Signed-off-by: Allen Bai <abai@redhat.com>

zonggen commented:
Source of truth for each part of data collected:
- Minimal:
  - platform: read_file `/proc/cmdline`
  - current_os_version: read_output `rpm-ostree status --json`
  - original_os_version: read_file `/.coreos-aleph-version.json`
  - instance_type: read_file `/run/metadata/afterburn`
- Full:
  - hardware:
    - disk: read_output `lsblk --fs --json`
    - cpu: read_output `lscpu --json`
    - memory: read_output `lsmem --json`
  - network: read_output `nmcli device show`

Commit:
src/full: add module for extracting container runtime info

Signed-off-by: Allen Bai <abai@redhat.com>

zonggen commented:
STATUS:
 - Since collecting service is meant to running in the background and users should not expect visible error information, the program should not crash/return error when one of the commands returns non-zero and instead should suppress the error and collect whatever is available at the moment.
 - Could use a log file to dump the error message

Commit:
agent/minimum: set default value for identity instead of returning error
tests: add tests to print minimum and full agent with test config file

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
main.rs: serialize the Agent struct and POST as json

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
*: add some documentation for each module
Dockerfile: install dependency, openssl-devel

Signed-off-by: Allen Bai <abai@redhat.com>

zonggen commented:
STATUS:
 - Client side data collection is done. though the `send_data()` function is not doing anything other than printing the collected data in Json format. This needs to be updated after we have an actual endpoint.
 - Persistent state so that pinger could send data periodically with some fixed interval will be added through another PR to reduce the size of this PR

Overall, ready for review!

Commit:
utl/clock: add clock module to record time intervals

Since reporting will happen daily or monthly, we are keeping
track of time intervals between reports. Two threads are spawned
for tracking daily and monthly reporting. Daily thread will check
timestamp every 12 hours, otherwise put to sleep. Monthly thread
will check every 15 days, otherwise put to sleep.

Signed-off-by: Allen Bai <abai@redhat.com>

zonggen commented:
STATUS:
 - Built on top of https://github.com/coreos/fedora-coreos-pinger/pull/35, in the main function, to create a Daemon process, one process is forked and parent process is immediately terminated. Two threads are then spawned by child process to track the time for daily and monthly reports.
 - The two threads are put into sleep to avoid busy waiting, and only checks whether it needs to send new report every 12 hours / 15 days for daily and monthly reports correspondingly.

zonggen commented:
cc @lucab @bgilbert

Commit:
full/container_runtime: rewrite to use `pgrep` instead of calling runtime

Previously we are using individual container runtimes to extract container
information, e.g. calling `podman container ls` to count running containers.
It would create two problems:
  - podman failed when trying to create `/.config/containers` since root dir
    is read-only and dynamic user does not have its own home dir
  - successful calls do not make sense either since counting running containers
    under the dynamic user is not useful
Hence, switch to count containers by calling `pgrep` and count the running
processes of each container runtime.

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
systemd: use `LogsDirecoty=` to persist data

Previously used `StateDirectory=` to persist pinger specific
data across bootups. It turns out that systemd does not allow
sharing state directory across different dynamic users. An example
would be:

```
$ systemd-run --pty --property=DynamicUser=yes --property=StateDirectory=wuff /bin/sh
$ systemd-run --pty --property=DynamicUser=yes --property=StateDirectory=wuff /bin/sh
```

Reference: http://0pointer.net/blog/dynamic-users-with-systemd.html
Signed-off-by: Allen Bai <abai@redhat.com>

ashcrow commented:
Is this something we want to continue pursuing?

bgilbert commented:
Maybe not; xref https://github.com/coreos/fedora-coreos-tracker/issues/770.
